include/deal.II/base/mpi.h

```diff
@@ -2026,6 +2053,80 @@ namespace Utilities
     }
 #  endif
 
+
+
+    template <typename T>
+    std::pair<T, T>
+    prefix_sum(const T &value, const MPI_Comm &comm, const bool exclusive)
+    {
+      std::pair<T, T> result;
+
+      prefix_sum<T>(ArrayView<const T>(&value, 1),
+                    comm,
+                    ArrayView<std::pair<T, T>>(&result, 1),
+                    exclusive);
+      return result;
+    }
+
+
+
+    template <typename T>
+    std::vector<std::pair<T, T>>
+    prefix_sum(const std::vector<T> &values,
+               const MPI_Comm &      comm,
+               const bool            exclusive)
+    {
+      const unsigned int           N = values.size();
+      std::vector<std::pair<T, T>> results(N);
+
+      prefix_sum<T>(ArrayView<const T>(values.data(), N),
+                    comm,
+                    ArrayView<std::pair<T, T>>(results.data(), N),
+                    exclusive);
+      return results;
+    }
+
+
+
+    template <typename T>
+    void
+    prefix_sum(const ArrayView<const T> &        values,
+               const MPI_Comm &                  comm,
+               const ArrayView<std::pair<T, T>> &results,
+               const bool                        exclusive)
+    {
+      AssertDimension(values.size(), results.size());
+
+      std::vector<T> temp(values.size());
+      std::vector<T> sums(values.size());
+
+      if (exclusive)
+        {
+          int ierr = MPI_Exscan(values.data(),
+                                temp.data(),
+                                values.size(),
+                                Utilities::MPI::mpi_type_id_for_type<T>,
+                                MPI_SUM,
+                                comm);
+          AssertThrowMPI(ierr);
+        }
+      else
+        {
+          int ierr = MPI_Scan(values.data(),
+                              temp.data(),
+                              values.size(),
+                              Utilities::MPI::mpi_type_id_for_type<T>,
+                              MPI_SUM,
+                              comm);
+          AssertThrowMPI(ierr);
+        }
+
+      Utilities::MPI::sum(values, comm, make_array_view(sums));
```

Comment:
Couldn't you omit this separate sum step in favor of:
* always using the `MPI_Scan`, 
* broadcasting the final value from the last process,
* if `exclusive` then subtract the value of the current process
?

Reply:
The difference between `Scan` and `ExScan` is that the former does
```
  \sum_{i=0}^{my_rank} values[i];
```
whereas the latter does
```
  \sum_{i=0}^{my_rank-1} values[i];
```
I however agree that we could always use the `Scan` version and subtract the local value (no need for a broadcast).

Reply:
Wait, how would you omit the broadcast? You need the value from the last process (the total sum) somehow.